torch/csrc/distributed/rpc/py_rref.cpp

```diff
@@ -109,7 +109,10 @@ PyRRef::PyRRef(const py::object& value, const py::object& type_hint)
         TypePtr elem_type = tryInferTypeWithTypeHint(value, type_hint);
         auto rref = RRefContext::getInstance().createOwnerRRef(elem_type);
         py::object copy(value); // increases refcount
```

Comment:
Hey @wanchaol Do you know when we do `std::move` in the original code below, what happens to the reference cnt on the `py::object`? 

And why do we need the copy here? The `jit::toIValue` below should have already inc the refcnt, no?

Reply:
When you `move` copy into `toIValue`, `toIValue` *may* take ownership of copy, and you would no longer have a lifetime guarantee through value. However, because you are calling `PyRRef` with a const reference to `value`, the general assumption is that this means that `value` will stay live for the entirety of the function.

Reply:
Compiler only synthesizes move constructor for a class that has all its non-static members movable (have move constructor). 
Since `py::handle` has a non-movable member, `PyObject *m_ptr` (https://github.com/pybind/pybind11/blob/master/include/pybind11/pytypes.h#L217), `py::handle` does not have a synthesized move constructor. 

When a class does not have a move constructor defined, but an RValue is passed to its constructor, the copy constructor is called in stead.
So these two calls are equivalent. The case converges to the one mentioned above.
```cpp
IValue ivalue = jit::toIValue(copy, elem_type);
IValue ivalue = jit::toIValue(std::move(copy), elem_type);
```